text/0000-feature-descriptions-doc-cfg.md

```diff
@@ -0,0 +1,208 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Usage is simple: features will be able to be specified in a table (inline or
+separate) with the keys `doc`, `public`, and `requires`. Sample `Cargo.toml`:
+
+```toml
+# Cargo.toml
+
+[features]
+# current configuration
+foo = []
+# Add a description to the feature
+bar = { requires = ["foo"], doc = "simple docstring here"}
+# `public` indicates whether or not the feature should be visible in
+# documentation, and defaults to true
+baz = { requires = ["foo"], public = false }
+
+# Features can also be full tables if descriptions are longer
+[features.qux]
+requires = ["bar", "baz"]
+doc = """
+# qux
+
+This could be a longer description of this feature
+"""
+```
+
+This RFC will also enable a `[tools.rustdoc]` table where existing configuration
+can be specified
+
+```toml
+# Cargo.toml
+
+[tools.rustdoc]
+html-logo-url = "https://example.com/logo.jpg"
+issue-tracker-base-url = "https://github.com/rust-lang/rust/issues/"
+```
+
+For projects that do not use cargo or want separate configuration, these options
+can also be specified in a `rustdoc.toml` file
+
+```toml
+# rustdoc.toml containing the same information as above
+
+html-logo-url = "https://example.com/logo.jpg"
+issue-tracker-base-url = "https://github.com/rust-lang/rust/issues/"
+
+[features]
+# current configuration
+foo = []
+# Add a description to the feature
+bar = { requires = ["foo"], doc = "simple docstring here"}
+
+# (baz and qux features clipped)
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+What exactly `rustdoc` does with the information is TBD. There are two options 
+
+## JSON Configuration
+
+`rustdoc` will gain a `--config-json` argument that allows passing a
+JSON-serialized string of the TOML configuration. It is likely that this is what
+Cargo can use when it invokes `rustdoc`: all that is needed is to parse the
+`features` and `tools.rustdoc` table from `Cargo.toml`, serialize to JSON, and
+pass as an argument.
+
+```sh
+rustdoc --argfoo --argbar . --config-json '{"html-logo-url":
+"https://example.com/logo.jpg","issue-tracker-base-url":
+"https://github.com/rust-lang/rust/issues/","features":{"foo":[],"bar":{"doc":
+"simple docstring here","requires":["foo"]},"baz":{"public":false,"requires":
+["foo"]},"qux":{"doc":"# qux\n\nThis could be a longer description of this feature\n"
+,"requires":["bar","baz"]}}}'
+```
+
+This sort of format has two distinct advantages:
+
+1. Build systems other than `cargo` can easily make use of the configuration,
+   and a `rustdoc.toml` file isn't required (e.g. if the build system has a
+   single configuration file for all tools)
+2. `rustdoc` does not need to be aware of `cargo`, paths, workspaces, etc.
+3. `rustdoc` can share the same `serde` structs to parse both `rustdoc.toml` or
+   this JSON configuration
+
+Arguments longer than the allowed limit (8000ish on Windows I think) can use
```

Comment:
I'm not the greatest fan of using argfiles for json data because that makes it harder to use with json dedicated processing tools like jq, firefox's json viewer, etc. Ideally one would have one flag for specifying (small) jsons on the command line, and one flag for specifying a file that is valid json instead of being surrounded by `--config-json '<valid json here>'`.

Also note that for error reporting purposes you probably want span info for all the stuff: say someone puts a code block or something which fails to test.

For coordination, RFC #3389 has to solve more or less the same tasks: sending potentially large amounts of data from Cargo.toml to rustc. Ideally that also would use valid json files instead of argfiles :).

Reply:
Good points with this, I have updated to disregard argfiles and added info about spans.

Thanks for the related issue - do you think anything there changes this RFC or vice versa? I did add at the end of future possibilities that maybe `rustfmt` or `clippy` could eventually use a similar design (which I would love to see)

Reply:
I don't think anything is blocking, but I will bring it up on that thread. This is now in the split RFC at https://github.com/rust-lang/rfcs/pull/3421